core/connection.go

```diff
@@ -0,0 +1,76 @@
+package connection
+
+import (
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+)
+
+type PermanentError struct {
+	Inner error
+}
+
+func (e PermanentError) Error() string { return e.Inner.Error() }
+func (e PermanentError) Unwrap() error {
+	return e.Inner
+}
+func (e PermanentError) Is(err error) bool {
+	_, ok := err.(PermanentError)
+	return ok
+}
+
+// Retries a given function in an exponential backoff manner and returns a value upon correct execution.
+// It will retry calling the function while it returns a non permanent error, until the max retries.
+// If maxTries == 0 then the retry function will run indefinitely until success or until a `PermanentError` is returned.
+func RetryWithData[T any](functionToRetry func() (*T, error), minDelay uint64, factor float64, maxTries uint64) (*T, error) {
```

Comment:
is it called "withdata" because it returns data? or data is related to some sort of parameter? a more clear name would be better